### PR TITLE
Auth beforefilter

### LINKED ISSUE #6905: RFC - Change AuthComponent::startup() to AuthComponent::beforeFilter()
At times i have seen devs trying to get logged in user info in controller's `beforeFilter()` when using stateless authenticators. That doesn't work since AuthComponent checks for user record availability through stateless authenticators from it's `startup()` callback which is run after controller's `beforeFilter()`.

So changing `AuthComponent::startup()` to `AuthComponent::beforeFilter()` would fix this issue as component's `beforeFilter()` callbacks are triggered before controller's `beforeFilter()`. The component can be configured through controller's `initialize()` so there's no real need delay the authentication.

Though this change would affect those setting up Auth in `beforeFilter()` and hence would not be backwards compatible.

### PR DESCRIPTION
**Pull request by ADmad**
Refs #6905

### CODE REVIEW

**Comment by ADmad** [comment]
Suggestions welcome for a better config name.

**Comment by dakota** [comment]
`legacyAuth` perhaps?

File: src/Controller/Component/AuthComponent.php [hunk 1]
```diff
@@ -147,7 +154,8 @@ class AuthComponent extends Component
         'logoutRedirect' => null,
         'authError' => null,
         'unauthorizedRedirect' => true,
-        'storage' => 'Session'
+        'storage' => 'Session',
+        'earlyAuth' => false
```
Comment (markstory) [inline src/Controller/Component/AuthComponent.php]: Would it be crazy to have an option named `authenticateIn` that takes `beforeFilter` or `startup` as values?

File: src/Controller/Component/AuthComponent.php [hunk 2]
```diff
@@ -281,18 +315,6 @@ public function startup(Event $event)
...
-     * Events supported by this component.
-     *
-     * @return array
-     */
-    public function implementedEvents()
-    {
-        return [
-            'Controller.startup' => 'startup',
-        ];
-    }
```
Comment (markstory) [inline src/Controller/Component/AuthComponent.php]: Having this be explicit avoids the reflection calls (micro-optimization).

File: src/Controller/Component/AuthComponent.php (hunk 1 above)
Comment (ADmad) [inline src/Controller/Component/AuthComponent.php]: Not crazy, I'll wait a bit before updating to allow others to weigh in.

File: src/Controller/Component/AuthComponent.php (hunk 2 above)
Comment (ADmad) [inline src/Controller/Component/AuthComponent.php]: Will revert and add beforeFilter to the list.

File: src/Controller/Component/AuthComponent.php (hunk 1 above)
Comment (jeremyharris) [inline src/Controller/Component/AuthComponent.php]: Would it be possible to pass an event string that it registers? That way it could be almost anything.

Comment (ADmad) [inline src/Controller/Component/AuthComponent.php]: @jeremyharris I don't think that would be easy as allowing any event would require using implementedEvents() and events get registered as soon as component instance is created before `Controller::initialize()` is called.

**Comment by ADmad** [comment]
@dakota Thanks but I went with Mark's suggestion :)

@markstory Addressed your comments.

@jeremyharris The auth check can now be triggered on any configured event. Just for custom event you will have to take care of attaching AuthComponent as listener for that event yourself.

**Comment by jeremyharris** [comment]
@ADmad oh cool, it could be done after all! Nice work!

**Comment by markstory** [comment]
Looks good outside of PHPCS errors.

**Comment by ADmad** [comment]
CS errors fixed.

**Pull request merged**